src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.SolutionSize;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// A service that enables storing and retrieving of information associated with solutions,
+    /// projects or documents across runtime sessions.
+    /// </summary>
+    internal abstract partial class AbstractPersistentStorageService : IPersistentStorageService2
+    {
+        private readonly IOptionService _optionService;
+        private readonly IPersistentStorageLocationService _locationService;
+        private readonly SolutionSizeTracker _solutionSizeTracker;
+
+        /// <summary>
+        /// This lock guards all mutable fields in this type.
+        /// </summary>
+        private readonly object _lock = new object();
+        private ReferenceCountedDisposable<IPersistentStorage> _currentPersistentStorage;
+        private SolutionId _currentPersistentStorageSolutionId;
```

Comment:
so it cache the last solution rather than current primary solution.

Reply:
Ah, right I can probably add a check that it's the one in the workspace.

Note: this is no longer a singleton but a proper workspace service, so there's one of these per workspace. Practically, there's only one SolutionId per workspace as well, although I do have a bug here thinking about it now where if you did have a change to the SolutionId and forgot to raise the storage location changed, I might fail to dispose the other one.